### PR TITLE
Update to Shopgate version 2.9.49

### PR DESCRIPTION
**Pull request by OliverMuellerShopgate**
Update to Shopgate version 2.9.49

### CODE REVIEW

File: classes/Customer.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * obtain it through the world-wide-web, please send an email
+ * to interfaces@shopgate.com so we can send you a copy immediately.
+ *
+ * @author    Shopgate GmbH, Schloßstraße 10, 35510 Butzbach <interfaces@shopgate.com>
+ * @copyright Shopgate GmbH
+ * @license   http://opensource.org/licenses/AFL-3.0 Academic Free License ("AFL"), in the version 3.0
+ */
+
+if (version_compare(_PS_VERSION_, '1.5.0', '<')) {
+    include_once(_PS_MODULE_DIR_.'shopgate'.DIRECTORY_SEPARATOR.'classes'.DIRECTORY_SEPARATOR.'ModObjectModelCustomer.php');
```
Comment (mmoustai) [inline classes/Customer.php]: You should use require_once rather than include_once. Generally, it's best to know immediately if something you need and expect isn't available.

File: classes/Customer.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+} else {
+    include_once(_PS_MODULE_DIR_.'shopgate'.DIRECTORY_SEPARATOR.'classes'.DIRECTORY_SEPARATOR.'ModObjectModelCustomerDummy.php');
```
Comment (mmoustai) [inline classes/Customer.php]: You can remove the else statement and directly extends your class like this

``` php
class ShopgateCustomerPrestashop extends ObjectModel
```

File: classes/Order.php [hunk 3]
```diff
@@ -156,4 +173,280 @@ public function updateFromOrder($order)
...
+                    //   value, cause we got it at this point
+                    if (empty($refundQtyData[$pItem['id_order_detail']]) && Tools::isSubmit('partialRefundProduct')) {
+                        continue;
+                    } else {
+                        $qty = $refundQtyData[$pItem['id_order_detail']];
+                    }
+                    
+                    if (version_compare(_PS_VERSION_, '1.5.0', '>=') && empty($qty)) {
+                        $qty = $sgItem->getQuantity() - $pItem['product_quantity'];
+                    } elseif (!empty($_POST['cancelQuantity'][$pItem['id_order_detail']]) && empty($qty)) {
```
Comment (mmoustai) [inline classes/Order.php]: Use Tools::getValue instead of $_POST

File: classes/Order.php [hunk 4]
```diff
@@ -156,4 +173,280 @@ public function updateFromOrder($order)
...
+                        $qty = $_POST['cancelQuantity'][$pItem['id_order_detail']];
```
Comment (mmoustai) [inline classes/Order.php]: Use Tools::getValue instead of $_POST

**Comment by OliverMuellerShopgate** [comment]
@Mekki78 Did you finish the review, so I can start to fix the issues you found?

File: classes/Order.php [hunk 5]
```diff
@@ -77,7 +94,7 @@ public function getFields()
         $fields['shipping_service'] = pSQL($this->shipping_service);
         $fields['shipping_cost']     = (float)($this->shipping_cost);
         $fields['shop_number']         = pSQL($this->shop_number);
-        $fields['comments']         = pSQL($this->comments, true);
+        $fields['comments']         = $this->comments;
```
Comment (Quetzacoalt91) [inline classes/Order.php]: Why did you remove the pSQL call ?

File: classes/items/order/export/Json.php [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            $order->setPaymentMethod($orderCore->payment);
+
+            $order->setIsShippingCompleted($orderCore->hasBeenShipped());
+            $order->setShippingCompletedTime($orderCore->hasBeenShipped() ? $orderCore->delivery_date : null);
+
+            $order->setDeliveryNotes($this->_getDeliveryNotes($orderCore));
+            $order->setExtraCosts($this->_getExtraCost($orderCore));
+            $order->setOrderTaxes($this->_getOrderTaxes());
+
+            array_push($this->_result, $order);
```
Comment (Quetzacoalt91) [inline classes/items/order/export/Json.php]: When you add only one element to the array, you should really consider to use 

``` php
$this->_result[] = $order;
```

I see that array_push everywhere and the other way is quicker to execute, because you do not make a function call.

File: shopgate.php [hunk 7]
```diff
@@ -171,6 +173,8 @@ public function install()
             'header',
             'adminOrder',
             'updateOrderStatus',
+            'updateQuantity',
+            'actionUpdateQuantity',
```
Comment (Quetzacoalt91) [inline shopgate.php]: You need only the first **updateQuantity** hook. On PrestaShop 1.5+, an alias exists in order to link the new and old name (See table hook_alias)

File: shopgate.php [hunk 8]
```diff
@@ -171,6 +173,8 @@ public function install()
             'header',
             'adminOrder',
             'updateOrderStatus',
+            'updateQuantity',
```
Comment (Quetzacoalt91) [inline shopgate.php:176]: Did you manage to register this new hook during an upgrade ?

**Comment by Quetzacoalt91** [comment]
Hi @OliverMuellerShopgate,

You're good to go now. :+1:

**Comment by mmoustai** [comment]
Warning after Save Button [BO]

download 3

**Comment by OliverMuellerShopgate** [comment]
@Mekki78 Its a permission issue. You need to set permissions to 774 for the folder modules/shopgate/vendors/shopgate_library/temp and all underlying folders. I also created a ticket for our instruction on support.shopgate.com.

File: classes/Customer.php (hunk 2 above)
Comment (OliverMuellerShopgate) [inline classes/Customer.php]: This won't give me the same result. I want to extend the ObjectModel in case its a Prestashop version <= 1.5.0.0 (see file ModObjectModelCustomer.php)

File: shopgate.php (hunk 7 above)
Comment (OliverMuellerShopgate) [inline shopgate.php]: The current code is tested and we had some curious behavior. The developer of this feature told me that the hook for "actionUpdateQuantity" was called twice (before updateQuantity hook was added to the code) in version 1.6.x.x thats why we added this condition "&& !empty($param['product']) && (!($param['product'] instanceof ProductCore))" to prevent the code executed twice. (This works because the params differ between each call)

We will change that in a later point. As I said the current code is working and tested.

File: shopgate.php (hunk 8 above)
Comment (OliverMuellerShopgate) [inline shopgate.php:176]: Do you have any documentation about the update process on modules of Prestashop? Is it "just" the function update() that needs to be implemented? I couldn't find something. Since what Prestashop version are you supporting the update of a module?

Comment (OliverMuellerShopgate) [inline shopgate.php:176]: @Quetzacoalt91 I found this documentation: http://doc.prestashop.com/display/PS15/Auto-updating+modules

Is this what you are talking about? How can I test if my implementation works? I think the "Update" only appears if my module is already in the market, right? So how to test it?

Comment (OliverMuellerShopgate) [inline shopgate.php:176]: @Quetzacoalt91 Am I correct, that the passed $module variable contains an instance of class object "ShopGate" (extends PaymentModule) or is this object a plain "PaymentModule" one?

File: shopgate.php (hunk 7 above)
Comment (Quetzacoalt91) [inline shopgate.php]: Interesting, these two hooks should NOT have different content in their param `$param`. In this case, your version is acceptable.

File: shopgate.php (hunk 8 above)
Comment (Quetzacoalt91) [inline shopgate.php:176]: Hi @OliverMuellerShopgate,

Exactly, the `$module` parameter will be an instance of ShopGate. :)
And yes, the documentation you found is what I was talking about. These upgrade are usefull when the module database is updated between two versions, or when hooks need to be registered or unregistered.

If you want to test your upgrade file, just install your module from the Backoffice (which will be downloaded from Addons), and then upload your upgraded version. If everything goes right, you will see at the top of the page that an upgrade file has been successfully applied.

**Comment by OliverMuellerShopgate** [comment]
@Quetzacoalt91 I made a lot of changes. Everything should be fine now. Please approve the pull request.

**Comment by Quetzacoalt91** [comment]
Hi @OliverMuellerShopgate,

Thanks, Your Has just been merged and the changes will be available from tomorrow.

Best regards

**Comment by OliverMuellerShopgate** [comment]
Hi @Quetzacoalt91,

thank you!

Best regards,

Oliver
